api/src/main/java/com/vimeo/networking2/VimeoResponse.kt

```diff
@@ -0,0 +1,51 @@
+package com.vimeo.networking2
+
+import com.vimeo.networking2.VimeoResponse.Error
+import com.vimeo.networking2.VimeoResponse.Success
+
+/**
+ * Result of the API response. [Success] contains the data while the [Error] class will
+ * inform you of an api error.
+ */
+sealed class VimeoResponse<out T> {
+
+    /**
+     * A successful response.
+     *
+     * @param data The parsed data for the request.
+     */
+    data class Success<out T>(val data: T) : VimeoResponse<T>()
+
+    /**
+     * An error occurred when making the request. This error may be due to invalid parameters,
+     * exception thrown when making the request or there was an error parsing the response.
+     *
+     * @param message The error message.
+     */
+    sealed class Error(val message: String): VimeoResponse<Nothing>() {
+
+        /**
+         * Vimeo API returned an error response for the request you made.
+         *
+         * @param reason Info on the error.
+         */
+        data class Api(val reason: ApiError): Error("API error")
+
+        /**
+         * Exception was thrown when making the request. This maybe due to no internet.
+         *
+         * @param throwable Info on the exception that was thrown.
+         */
+        data class Exception(val throwable: Throwable): Error("Exception thrown")
+
+        /**
+         * Parsing error occurred. The request was successful, but the response could not be
+         * parsed by the SDK.
+         *
+         * @param rawResponse Raw response from the API.
+         */
+        data class Parsing(val rawResponse: String): Error("Parsing error")
```

Comment:
It seems like this error replaces the generic error - what is the logic behind any error that doesn't have a `reason` being a parsing error? What if a request is made for a missing resource and we get a 404? Will that be surfaced as a `Parsing` error, and, if so, is that really the "right" error?

Reply:
A 404 error gives the following response from the API: 

```
{
  "error": "The requested page could not be found"
}
```

If this error is successfully parsed, it will be surfaced to the client as an API error as follow:

```
ApiError(
    developerMessage=null, 
     errorMessage=The requested page could not be found, 
     errorCode=null, 
     invalidParameters=null, link=null
)
```

Suppose this error occurs and the error convert fails to convert it into a `ApiError` object, the client will be informed that a parsing error occurred. This will be the response to the client:

```
Parsing(
   rawResponse=Response{  protocol=http/1.1, 
                                                code=404, 
                                                message=Not Found, 
                                                url=https://api.vimeo.com/oauth/authorize/clientttttt
   })
```

I could rename `Parsing` to `Generic` for a better name.